File: 2021/10/06/shortest_to_char.py
"""
https://leetcode.com/problems/shortest-distance-to-a-character/
Author: Andrew Jarombek
Date: 10/6/2021
"""

from typing import List


def shortest_to_char(s: str, c: str) -> List[int]:
    result = [len(s) for _ in range(len(s))]

    for index, char in enumerate(s):
        if char == c:
            result[index] = 0

            distance = 1
            for i in range(index - 1, -1, -1):
                if result[i] > distance:
                    result[i] = distance
                    distance += 1
                else:
                    break

            distance = 1
            for i in range(index + 1, len(s)):
                if result[i] > distance:
                    result[i] = distance
                    distance += 1
                else:
                    break

    return result


if __name__ == '__main__':
    assert shortest_to_char('aaab', 'b') == [3, 2, 1, 0]

File: 2023/01/05/shortestToChar.js
/**
 * https://leetcode.com/problems/shortest-distance-to-a-character/
 * @author Andrew Jarombek
 * @since 1/5/2023
 */

const shortestToChar = function(s, c) {
    const locations = [-Infinity];

    for (let i = 0; i < s.length; i++) {
        if (s.charAt(i) === c) {
            locations.push(i);
        }
    }

    locations.push(Infinity);
    const result = [];
    let index = 0;

    for (let i = 0; i < s.length; i++) {
        const val = Math.min(i - locations[index], locations[index+1] -i);
        result.push(val);

        if (val === 0) {
            index++;
        }
    }

    return result;
}

const result = shortestToChar("aaab", "b");
const expected = [3,2,1,0]

console.assert(result.every((value, index) => value === expected[index]));
